openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -117,6 +111,304 @@ class NonbondedMethod(Enum):
     PME = 4
 
 
+#======================================================================
+# PARAMETER ATTRIBUTES
+#======================================================================
+
+# TODO: Think about adding attrs to the dependencies and inherit from attr.ib
+class ParameterAttribute:
+    """A descriptor for ``ParameterType`` attributes.
+
+    The descriptors allows associating to the parameter a default value,
+    which makes the attribute optional, a unit, and a custom converter.
+
+    Because we may want to have ``None`` as a default value, required
+    attributes have the ``default`` set to the special type ``UNDEFINED``.
+
+    Converters can be both static or instance functions/methods with
+    respective signatures
+
+    converter(value): -> converted_value
+    converter(instance, parameter_attribute, value): -> converted_value
+
+    A decorator syntax is available (see example below).
+
+    Parameters
+    ----------
+    default : object, optional
+        When specified, the descriptor makes this attribute optional by
+        attaching a default value to it.
+    unit : simtk.unit.Quantity, optional
+        When specified, only quantities with compatible units are allowed
+        to be set, and string expressions are automatically parsed into a
+        ``Quantity``.
+    converter : callable, optional
+        An optional function that can be used to convert values before
+        setting the attribute.
+
+    Examples
+    -------
+
+    Create a parameter type with an optional and a required attributes.
+
+    >>> class MyParameter:
+    ...     attr_required = ParameterAttribute()
+    ...     attr_optional = ParameterAttribute(default=2)
+    ...
+    >>> my_par = MyParameter()
```

Comment:
Ahhh, I just finished the rest of the review, and now I see that required `ParameterAttribute`s are checked by the `ParameterType` that owns them. Since `class MyParameter` doesn't inherit from `ParameterType`, `attr_required` isn't really required. 

The docstring does become a bit confusing when we use the word "required" like this. Maybe we could change it to something like `normal_attr = ParameterAttribute()` `attr_with_default = ParameterAttribute(default=2)`?

Reply:
Yeah, I don't think there's a way for descriptors to verify they've been set when the class that owns them is initialized. I'll fix the examples.

Reply:
I've just realized that I'll have to encapsulate the initialization utilities in a separate class/decorator shared between `ParameterType` and `ParameterHandler` so I'll fix it by making `MyParameter` using that class.